False-positive TS2683 for `this` in directly yielded Effect.gen expression

Hi! We may have found a false-positive `TS2683` when using `this` inside a directly yielded expression in `Effect.gen({ self })`.

Minimal repro:

```ts
import { Effect, Exit, Scope } from "effect";

class Repro {
  readonly #scope: Scope.Closeable = Scope.makeUnsafe();

  run(): Effect.Effect<void> {
    return Effect.gen({ self: this }, function* () {
      yield* Scope.close(this.#scope, Exit.void);
    });
  }
}
```

With `@effect/language-service` enabled, this reports:

```text
repro.ts(8,26): error TS2683: 'this' implicitly has type 'any' because it does not have a type annotation.
```

Equivalent code without `this` in the directly yielded expression passes:

```ts
run(): Effect.Effect<void> {
  return Effect.gen({ self: this }, function* () {
    const scope = this.#scope;
    yield* Scope.close(scope, Exit.void);
  });
}
```

I checked this against:

- `tsc`: passes
- `tsgo` without `@effect/language-service`: passes
- `tsgo` with `@effect/language-service`: reports `TS2683`
- runtime behavior: both code shapes behave the same in our test

So our current conclusion is that the plugin may be losing the contextual `this: Self` from the `Effect.gen({ self })` overload while analyzing the directly yielded expression. But we may be missing an intended constraint, so please let us know if this is expected.

Versions:

```text
@effect/tsgo: 0.5.1
tsgo: Version 7.0.0-dev+effect-tsgo.0.5.1
@typescript/native-preview: 7.0.0-dev.20260426.1
effect: 4.0.0-beta.57
typescript: 6.0.3
node: v25.9.0
bun: 1.3.11
```

Thanks!

<sub>Drafted by GPT-5.5 and reviewed by a human before submission.</sub>

--- a/.changeset/fix-issue-173-this-signature-help.md
+++ b/.changeset/fix-issue-173-this-signature-help.md
@@ -1,0 +1,7 @@
+---
+"@effect/tsgo": patch
+---
+
+Fix a false-positive `TS2683` when using `this` inside directly yielded expressions in `Effect.gen({ self: this })`.
+
+This avoids losing contextual `this` typing during data-first call analysis for Effect generator code such as `yield* Scope.close(this.#scope, Exit.void)`.

--- a/internal/effecttest/issue_173_test.go
+++ b/internal/effecttest/issue_173_test.go
@@ -1,0 +1,84 @@
+package effecttest_test
+
+import (
+	"testing"
+
+	"github.com/microsoft/typescript-go/shim/fourslash"
+
+	_ "github.com/effect-ts/tsgo/etscheckerhooks"
+	_ "github.com/effect-ts/tsgo/etslshooks"
+	_ "github.com/effect-ts/tsgo/etstesthooks"
+)
+
+func TestIssue173_DirectYieldedThisDoesNotReportFalsePositiveTS2683(t *testing.T) {
+	t.Parallel()
+
+	const content = `// @Filename: /tsconfig.json
+{
+  "compilerOptions": {
+    "strict": true,
+    "target": "ESNext",
+    "module": "NodeNext",
+    "moduleResolution": "NodeNext",
+    "plugins": [
+      {
+        "name": "@effect/language-service"
+      }
+    ]
+  }
+}
+// @Filename: /repro.ts
+import { Effect, Exit, Scope } from "effect"
+
+class Repro {
+  readonly #scope: Scope.Closeable = Scope.makeUnsafe()
+
+  run(): Effect.Effect<void> {
+    return Effect.gen({ self: this }, function* () {
+      yield* Scope.close([|this|].#scope, Exit.void)
+    })
+  }
+}`
+
+	f, done := fourslash.NewFourslash(t, nil /*capabilities*/, content)
+	defer done()
+
+	f.VerifyNonSuggestionDiagnostics(t, nil)
+}
+
+func TestIssue173_HoistedThisScopeDoesNotReportTS2683(t *testing.T) {
+	t.Parallel()
+
+	const content = `// @Filename: /tsconfig.json
+{
+  "compilerOptions": {
+    "strict": true,
+    "target": "ESNext",
+    "module": "NodeNext",
+    "moduleResolution": "NodeNext",
+    "plugins": [
+      {
+        "name": "@effect/language-service"
+      }
+    ]
+  }
+}
+// @Filename: /repro.ts
+import { Effect, Exit, Scope } from "effect"
+
+class Repro {
+  readonly #scope: Scope.Closeable = Scope.makeUnsafe()
+
+  run(): Effect.Effect<void> {
+    return Effect.gen({ self: this }, function* () {
+      const scope = this.#scope
+      yield* Scope.close(scope, Exit.void)
+    })
+  }
+}`
+
+	f, done := fourslash.NewFourslash(t, nil /*capabilities*/, content)
+	defer done()
+
+	f.VerifyNonSuggestionDiagnostics(t, nil)
+}

--- a/internal/typeparser/data_first_signature.go
+++ b/internal/typeparser/data_first_signature.go
@@ -39,6 +39,9 @@ func (tp *TypeParser) DataFirstOrLastCall(node *ast.Node) *ParsedDataFirstOrLast
 	if len(resolved.Parameters()) != len(call.Arguments.Nodes) {
 		return nil
 	}
+	if tp.GetEffectContextFlags(node)&EffectContextFlagCanYieldEffect != 0 && callHasNonBareThisArgument(call.Arguments.Nodes) {
+		return nil
+	}
 
 	resolvedSymbol := checker.Checker_getSymbolOfDeclaration(c, resolved.Declaration())
 	if resolvedSymbol == nil {
@@ -104,6 +107,36 @@ func (tp *TypeParser) DataFirstOrLastCall(node *ast.Node) *ParsedDataFirstOrLast
 	}
 
 	return nil
+}
+
+func callHasNonBareThisArgument(args []*ast.Node) bool {
+	for _, arg := range args {
+		if arg == nil || arg.Kind == ast.KindThisKeyword {
+			continue
+		}
+		if containsThisKeyword(arg) {
+			return true
+		}
+	}
+	return false
+}
+
+func containsThisKeyword(node *ast.Node) bool {
+	if node == nil {
+		return false
+	}
+	if node.Kind == ast.KindThisKeyword {
+		return true
+	}
+	contains := false
+	node.ForEachChild(func(child *ast.Node) bool {
+		if containsThisKeyword(child) {
+			contains = true
+			return true
+		}
+		return false
+	})
+	return contains
 }
 
 func derivePipeableSignatureFromDataFirst(c *checker.Checker, sig *checker.Signature, subjectIndex int) *checker.Signature {

